From 927297f64c451c301cf99855ab7fe14729b97a1c Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Thu, 5 Sep 2024 16:04:29 +0530
Subject: [PATCH] wifi: ath12k: fix WARN_ON during rmmod ath12k

Some architectures don't use mon_reap_timer, so mon_reap_timer are not initialized
for such architectures. But in rmmod path try to delete mon_reap_timer,
due to this below warn_on triggered:

WARNING: CPU: 1 PID: 18565 at lib/debugobjects.c:528 debug_print_object+0x94/0xc4
ODEBUG: assert_init not available (active state 0) object type: timer_list hint: 0x0

(unwind_backtrace) from [<4010a1c4>] (show_stack+0x10/0x14)
(show_stack) from [<408464dc>] (dump_stack_lvl+0x40/0x4c)
(dump_stack_lvl) from [<4011b174>] (__warn+0x64/0xbc)
(__warn) from [<4011b2c8>] (warn_slowpath_fmt+0xfc/0x174)
(warn_slowpath_fmt) from [<40437984>] (debug_print_object+0x94/0xc4)
(debug_print_object) from [<40438734>] (debug_object_assert_init+0xc8/0x154)
(debug_object_assert_init) from [<40192ca4>] (try_to_del_timer_sync+0x20/0x90)
(try_to_del_timer_sync) from [<40192d50>] (del_timer_sync+0x3c/0x4c)
(del_timer_sync) from [<625fb680>] (ath12k_dp_pdev_free+0x24/0x70 [ath12k])
(ath12k_dp_pdev_free [ath12k]) from [<625bb888>] (ath12k_core_deinit+0x168/0x318 [ath12k])
(ath12k_core_deinit [ath12k]) from [<6264ebfc>] (ath12k_ahb_remove+0x4c/0x50 [ath12k])
(ath12k_ahb_remove [ath12k]) from [<404fb940>] (platform_remove+0x40/0x5c)
(platform_remove) from [<404f9d84>] (device_release_driver_internal+0x8c/0x148)
(device_release_driver_internal) from [<404f9ec8>] (driver_detach+0x70/0x7c)
(driver_detach) from [<404f858c>] (bus_remove_driver+0x88/0xb8)
(bus_remove_driver) from [<62657b3c>] (cleanup_module+0x2c/0x44 [ath12k])
(cleanup_module [ath12k]) from [<4018f1ac>] (sys_delete_module+0x170/0x29c)
(sys_delete_module) from [<40100060>] (ret_fast_syscall+0x0/0x4c)

To resolve this warning, avoid accessing to mon_reap_timer.

Patch-dependency: none

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c    | 8 +++++---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 9 +++++++--
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 4f577c3..3488aeb 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1401,7 +1401,8 @@ void ath12k_dp_pdev_free(struct ath12k_base *ab)
 	struct ath12k *ar;
 	int i;
 
-	del_timer_sync(&ab->mon_reap_timer);
+	if (!ab->hw_params->rxdma1_enable)
+		del_timer_sync(&ab->mon_reap_timer);
 
 	for (i = 0; i < ab->num_radios; i++) {
 		ar = ab->pdevs[i].ar;
@@ -1438,8 +1439,9 @@ static void ath12k_dp_service_mon_ring(struct timer_list *t)
 		ath12k_dp_mon_process_ring(ab, i, NULL, DP_MON_SERVICE_BUDGET,
 					   ATH12K_DP_RX_MONITOR_MODE);
 
-	mod_timer(&ab->mon_reap_timer, jiffies +
-		  msecs_to_jiffies(ATH12K_MON_TIMER_INTERVAL));
+	if (!ab->hw_params->rxdma1_enable)
+		mod_timer(&ab->mon_reap_timer, jiffies +
+			  msecs_to_jiffies(ATH12K_MON_TIMER_INTERVAL));
 }
 
 static void ath12k_dp_mon_reap_timer_init(struct ath12k_base *ab)
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 1cf023b..bc3f9d9 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -5772,6 +5772,9 @@ void ath12k_dp_rx_pdev_mon_detach(struct ath12k_base *ab, const int pdev_idx)
 
 int ath12k_dp_rx_pktlog_start(struct ath12k_base *ab)
 {
+	if (ab->hw_params->rxdma1_enable)
+		return 0;
+
 	/* start reap timer */
 	mod_timer(&ab->mon_reap_timer,
 		  jiffies + msecs_to_jiffies(ATH12K_MON_TIMER_INTERVAL));
@@ -5783,8 +5786,10 @@ int ath12k_dp_rx_pktlog_stop(struct ath12k_base *ab, bool stop_timer)
 {
 	int ret;
 
-	if (stop_timer)
-		del_timer_sync(&ab->mon_reap_timer);
+	if (!ab->hw_params->rxdma1_enable) {
+		if (stop_timer)
+			del_timer_sync(&ab->mon_reap_timer);
+	}
 
 	/* reap all the monitor related rings */
 	ret = ath12k_dp_purge_mon_ring(ab);
-- 
2.34.1

